test/ConnectionTest3.cpp

```diff
@@ -0,0 +1,52 @@
+#include <gtest/gtest.h>
+
+#include "DBConnectionCaster.hpp"
+
+namespace {
+
+    int getMySqlConnection3Info(const IDBConnection& dbConnection) {
+        auto& mysql_connection = DBConnectionCaster<MySqlDBConnection>::cast(dbConnection);
+        return mysql_connection.query();
+    }
+
+    std::string getMySqlConnection3AdvancedInfo(const IDBConnection& dbConnection) {
+        auto& mysql_connection = DBConnectionCaster<MySqlDBConnection>::cast(dbConnection);
+        return mysql_connection.advancedQuery().server_version;
+    }
+
+    int getSqLiteConnection3Info(const IDBConnection& dbConnection) {
+        auto& sqlite_connection = DBConnectionCaster<SqLiteDBConnection>::cast(dbConnection);
+        return sqlite_connection.query();
+    }
+
+    TEST(MySqlConnection3InfoTest, GetInfoStaticCast) {
+        MySqlDBConnection mysql_connection("Some server", 7);
+        ASSERT_EQ(7, getMySqlConnection3Info(mysql_connection));
+    }
+
+    TEST(MySqlConnection3InfoTest, GetAdvancedInfoStaticCast) {
+        MySqlDBConnection mysql_connection("One more server", 7);
+        ASSERT_EQ("One more server", getMySqlConnection3AdvancedInfo(mysql_connection));
+    }
+
+    TEST(SqLiteConnection3InfoTest, GetInfoStaticCast) {
+        SqLiteDBConnection sqlite_connection(3);
+        ASSERT_EQ(3, getMySqlConnection3Info(sqlite_connection));
+    }
+
+    TEST(SqLiteConnection3InfoTest, GetAdvancedInfoStaticCast) {
+        SqLiteDBConnection sqlite_connection(3);
+        ASSERT_ANY_THROW(getMySqlConnection3AdvancedInfo(sqlite_connection));
```

Comment:
Why not catching specific exception like its done below?

Reply:
@malirod , because we have different exceptions in debug and release modes. (debug - std::bad_alloc; release - std::length_error). So, to catch them we need to add `#ifndef NDEBUG` directives

Reply:
So why in the another case below we have single case?

Reply:
@malirod , in this case we have successful static_cast, but when `mysql_connection.advancedQuery().server_version` used - we have invalid memory reference. In case below we have `std::bad_cast` in release and debug modes

Reply:
This is sufficient explanation for. You have ub here and the whole point was to avoid it.